File: 77-combinations/combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> sol;
        vector<int> nums;

        auto bkt= [&](auto&& self,int start){
            if(nums.size()==k){
                sol.push_back(nums);
                return;
            }

            for(int i=start;i<=n;++i){
                nums.push_back(i);
                self(self,i+1);
                nums.pop_back();
            }
        };

        bkt(bkt,1);
        return sol;
    }
};

File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        sol = []
        temp = []

        def bkt(n,k,curIdx):
            if len(temp)==k:
                sol.append(temp[:])
                return
            
            for i in range(curIdx+1,n):
                temp.append(i+1)
                bkt(n,k,i)
                temp.pop()
        
        bkt(n,k,-1)

        return sol